crates/turbo-tasks-memory/src/cell.rs

```diff
@@ -1,37 +1,291 @@
-use std::fmt::Debug;
+use std::{fmt::Debug, mem::take};
 
 use auto_hash_map::AutoSet;
-use turbo_tasks::{backend::CellContent, TaskId, TurboTasksBackendApi};
+use turbo_tasks::{
+    backend::CellContent,
+    event::{Event, EventListener},
+    TaskId, TurboTasksBackendApi,
+};
+
+#[derive(Debug)]
+pub(crate) enum FullCell {
+    UpdatedValue {
+        content: CellContent,
+        updates: u32,
+        dependent_tasks: AutoSet<TaskId>,
+    },
+    /// Someone wanted to read the content and it was not available. The content
+    /// is now being recomputed.
+    /// This is used when there are dependent tasks
+    Recomputing {
+        event: Event,
+        updates: u32,
+        dependent_tasks: AutoSet<TaskId>,
+    },
+}
 
 #[derive(Default, Debug)]
-pub struct Cell {
-    content: CellContent,
-    updates: u32,
-    pub(crate) dependent_tasks: AutoSet<TaskId>,
+pub(crate) enum Cell {
+    /// No content has been set yet, or it was removed for memory pressure
+    /// reasons.
+    #[default]
+    Empty,
+    /// Someone wanted to read the content and it was not available. The content
+    /// is now being recomputed.
+    /// This is only used when there are no dependent tasks
+    Recomputing {
+        event: Event,
+        updates: u32,
+    },
+    /// The content was set only once and is tracked.
+    InitialValue {
+        content: CellContent,
+        dependent_tasks: AutoSet<TaskId>,
+    },
+    // This is in a box so we don't need the updates counter for most cells that are only written
+    // once.
+    Full(Box<FullCell>),
+}
+
+#[derive(Debug)]
+pub struct RecomputingCell {
+    pub listener: EventListener,
+    pub schedule: bool,
 }
 
 impl Cell {
-    pub fn read_content(&mut self, reader: TaskId) -> CellContent {
-        self.dependent_tasks.insert(reader);
-        self.read_content_untracked()
+    pub fn remove_dependent_task(&mut self, task: TaskId) {
+        match self {
+            Cell::Empty
+            | Cell::Recomputing { .. }
+            | Cell::Full(box FullCell::Recomputing { .. }) => {}
```

Comment:
Shouldn't we remove the task from `Recomputing::dependent_tasks`? Or, add a comment explaining why not.